### PR TITLE
Pr compound

### PR DESCRIPTION
**Pull request by chkngit**
adding inter-inter compound mode: 
on top of Average compound mode, we add Distance based, Difference based, and Wedge compound modes
bd rate gain of ~2.3% with speed loss of ~30% (when MD classes are on)

### CODE REVIEW

**Comment by chkngit** [comment]
@joelsole @NaderMahdi  @mafonsoN   can we have feedback on this ASAP ?  we would like to merge this to master in order to start more important PRs assuming Compound to be integrated

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 1]
```diff
@@ -382,6 +386,9 @@ typedef struct ConvolveParams
     int32_t use_jnt_comp_avg;
     int32_t fwd_offset;
     int32_t bck_offset;
+#if COMP_MODE
+    int use_dist_wtd_comp_avg;
+#endif
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbDefinitions.h]: - 1 line

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 2]
```diff
@@ -76,7 +76,83 @@ enum COEFF_SCAN_TYPE
...
+    }
+    if (left_mi != NULL) {
+        if (has_second_ref(left_mi))
+            left_ctx = left_mi->compound_idx;
+        else if (left_mi->ref_frame[0] == ALTREF_FRAME)
+            left_ctx = 1;
+    }
+    return above_ctx + left_ctx + 3 * offset;
+}
+int get_comp_group_idx_context_enc(const MacroBlockD *xd) {
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbEntropyCoding.c]: Possibly add a line between the functions in this file

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 3]
```diff
@@ -1162,14 +1230,1793 @@ void svt_highbd_inter_predictor(const uint16_t *src, int32_t src_stride,
...
+                //   If sign requested is 0, we need to return index [0] the master
+                //   if sign requested is 1, we need to return the complement index [1]
+                //   instead.
+                wedge_params.signflip[w] = (avg < 32);
+            }
+        }
+    }
+}
+#endif  // !USE_PRECOMPUTED_WEDGE_SIGN
+static INLINE int get_wedge_bits_lookup(BlockSize sb_type) {
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: Add a line between functions for readability?

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 4]
```diff
@@ -92,6 +99,77 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+        );
+
+        candidatePtr->interinter_comp.wedge_index = candidatePtr->interinter_comp.wedge_index;
+        candidatePtr->interinter_comp.wedge_sign = candidatePtr->interinter_comp.wedge_sign;
+    }
+    else {
+        printf("ERROR: not used comp type\n");
+    }
+
+
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbModeDecision.c]: Remove extra lines

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 1 above)
Comment (chkngit) [reply Source/Lib/Common/Codec/EbDefinitions.h]: ok

File: Source/Lib/Common/Codec/EbEntropyCoding.c (hunk 2 above)
Comment (chkngit) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: ok

File: Source/Lib/Common/Codec/EbModeDecision.c (hunk 4 above)
Comment (chkngit) [reply Source/Lib/Common/Codec/EbModeDecision.c]: ok

**Pull request merged**